DOP-3168: Set up client side routing and basic application shell

drew-beckmen opened the pull request:
### Ticket

[DOP-3168](https://jira.mongodb.org/browse/DOP-3168)

### Description

- Adds a basic application shell with the LG Side Nav component
- adds `gatsby-browser` and `gatsby-ssr` to render the application wrapped in a `Layout` component. The application will conditionally render the side nav based on the current URL. As long as we navigate internally within the application, the navigation component will not remount, and it should therefore not send additional requests to the backend to fetch a user's entitled repos.
- Side nav is populated with the current user's entitled repositories.
- Client side routing is achieved with the Gatsby [File System API](https://www.gatsbyjs.com/docs/reference/routing/file-system-route-api/#syntax-client-only-routes)
- Update `tsconfig` so Jest works correctly with Typescript
- Updates Jest config to account for mapping between Gatsby's custom reach/router and generic @reach/router import
- Mocks Gatsby Link component, which would otherwise cause errors when testing components that use links like SideNav

### Screenshot 

<img width="1716" alt="Screen Shot 2022-08-04 at 11 19 11 AM" src="https://user-images.githubusercontent.com/17502635/182884335-b18e49df-7805-41bc-a94e-aaa964c6f762.png">

drew-beckmen commented on frontend/src/pages/repos/[repoName].tsx:
```diff
@@ -0,0 +1,19 @@
+
+function RepoPage({ params }: RepoProps) {
+  return (
+    <AppShell>
```
I don't love how this is rendered in both `index.tsx` and `[repoName].tsx`, but based on the designs, we'll have pages such as `pages/repos/[repoName]/versions/[versionId]` where we don't want to display the side nav. Does anyone know of a better pattern for this?

mmeigs replied:
Co-sign with Raymund here. Interested in watching `wrapPageElement` work.
Thought experiment for the hell of it - probably not worth the trade-offs if I'm understanding correctly - but could there be an argument for `versions` to be its own page/route?

drew-beckmen replied:
Great call @rayangler and @mmeigs! I have used the `wrapPageElement` function in my original PR at first, but later deleted it. I followed a similar pattern to the SO post above, and it seems to be working pretty well. If the user navigates between pages that all display the side nav, we don't send additional requests to the backend (I think this is actually due to Gatsby's `Link` component ). However, if the user goes to a page directly or refreshes the browser, React triggers a re-render which sends a request to the backend.

drew-beckmen commented on frontend/tsconfig.json:
```diff
@@ -33,8 +33,13 @@
+      "node_modules/@types"
+    ] /* Specify multiple folders that act like `./node_modules/@types`. */,
+    "types": [
+      "jest",
```
needed for Jest to work with TS in VSCode IDE

drew-beckmen replied:
I'm actually still having issues with Intellisence complaining in VSCode... Anyone else having this issue?

drew-beckmen replied:
To be more specific, I've isolated the issue to the monorepo setup. If I open just the `frontend` folder in VSCode, the IDE understands and provides completion for Jest tests written in the `tests` directory. However, if I open the entire `docs-hapley` directory in VSCode, VSCode can't properly source Jest. For example:
![Screen Shot 2022-08-08 at 8 01 52 PM](https://user-images.githubusercontent.com/17502635/183534626-fa0cca4c-684c-4de8-941e-67d3ca4146fa.png)

rayangler replied:
Nice find regarding the monorepo setup. I've had weird Typescript errors appear/disappear locally as well (although not from the test files)

drew-beckmen replied:
I've found that restarting the TypeScript server in VS Code can often do the trick.

rayangler reviewed (commented):
Just a quick look rn. Will take a closer look later

drew-beckmen commented on frontend/gatsby-node.js:
```diff
@@ -0,0 +1,20 @@
+const webpack = require('webpack');
+
+// Resolves a know issue with incompatibility between Emotion and Webpack 5
```
Based on [this slack thread](https://mongodb.slack.com/archives/CM90JQHEJ/p1645813751369449?thread_ts=1645727542.936729&cid=CM90JQHEJ), it seems the underlying issue is with Emotion.

mmeigs reviewed (approved):
LGTM :)

rayangler commented on frontend/src/components/Layout.tsx:
```diff
@@ -0,0 +1,39 @@
+  display: grid;
+  grid-template-areas: 'sidenav contents';
+  grid-template-columns: auto 1fr;
+  grid-template-rows: auto 1fr;
```
(Super nit) I think we can remove rows for now since I don't think we actually have rows in the grid with this layout

rayangler commented on frontend/src/components/ContentContainer.tsx:
```diff
@@ -0,0 +1,30 @@
+      />
+      <main
+        css={css`
+          margin: 72px 72px 0 72px;
```
This might be a big assumption, but it might make sense to just make all of the margins `72px`. Otherwise, we risk having content touching the edge of the bottom

drew-beckmen replied:
That sounds fair to me!

rayangler reviewed (commented):
Looks good overall. Just wanted to flag a couple of things

drew-beckmen commented on frontend/package.json:
```diff
@@ -34,19 +34,29 @@
-    "@types/jest": "^28.1.4",
+    "@types/jest": "^28.1.6",
     "@types/node": "^17.0.45",
+    "@types/reach__router": "^1.3.10",
```
Good call. I don't think it's necessary anymore, but if someone ever does leverage elements of the Reach Router API that ships with Gatsby, we'll need to add the package back in to satisfy TypeScript. I'll remove it for now.

rayangler reviewed (commented):
Not sure why this is happening but there's a margin on the side nav title that's causing the top border line below it from extending the full width of the side nav. Is this showing up locally for you as well? Not sure if it's just a rendering issue on my end or something

Margin:
<img width="308" alt="Screen Shot 2022-08-04 at 1 36 33 PM" src="https://user-images.githubusercontent.com/27821750/182915983-0f386dd5-6c5d-47de-a491-c6345ea3a8cf.png">

Border:

<img width="308" alt="Screen Shot 2022-08-04 at 1 44 30 PM" src="https://user-images.githubusercontent.com/27821750/182916375-1e8bd159-81ea-44d4-884f-2fb0d97a8fdb.png">

mmeigs commented:
> 

Looks like it's the max-width property that's stopping it.

drew-beckmen commented on frontend/src/components/Layout.tsx:
```diff
@@ -0,0 +1,39 @@
+  grid-template-areas: 'sidenav contents';
+  grid-template-columns: auto 1fr;
+  grid-template-rows: auto 1fr;
+`;
```
I just pushed a change that resolves this issue. Thanks for pointing out the LG recommendation — that was super helpful! Now, if the page content overflows the viewport but the nav bar does not, only the page content will be scrollable. If the nav bar overflows 100vh, it also becomes scrollable. This prevents the "collapsible" icon from being pushed outside the viewport.

drew-beckmen commented:
> > 
> 
> Looks like it's the max-width property that's stopping it.

Nice catch @rayangler and yep @mmeigs is 100% correct. We actually don't need that max-width CSS property at all and should instead be setting the padding such that the width resolves itself given the width of the parent container (the side nav). I've removed the `max-width` and adjusted the padding such that the spacing is maintained and the border extends to the end of the side nav. Thanks all for catching!

rayangler reviewed (commented):
Just wanted to flag one edge case and one question about a specific pattern/decision!

Also, would it be a good idea to have a unit test for the `SideNav` component with some mocked data in place of the fetch? Since we have some logic to determine what's being rendered there, it might be nice to have a basic test to ensure everything is being rendered the way we want it to. This can be handled separately, if we want to merge this PR in ASAP to unblock anything else.

drew-beckmen commented on frontend/src/pages/repos/[repoName].tsx:
```diff
@@ -0,0 +1,14 @@
+}
+
+function RepoPage({ params }: RepoProps) {
+  return <Repo repoName={params.repoName} />;
```
My primary reason for mapping the pages to components directly was because it allows us to avoid the messiness of routing logic in the test environment. When I made my original attempt at client side routing by serving the app out of a single page, testing became nightmarish because the `reach/router` that ships with Gatsby is a modified fork of the `reach/router` package that is generally available. Since Gatsby is not available in the test environment, I had to install the regular `reach/router` in devDependencies, which introduced a host of other issues with inconsistencies in version requirements.

rayangler replied:
Gotcha, using it for testing makes sense. Thanks!

drew-beckmen commented on frontend/src/components/SideNav.tsx:
```diff
@@ -0,0 +1,65 @@
+          const repoName: string = repo.split('/')[1];
+          return (
+            <SideNavItem
+              className={cx(navItemStyle(location.pathname.includes(repoName)))}
```
Nice catch! I've updated the logic to better handle these edge cases:
```
  const isActive = (repoName: string): boolean => {
    const currentDocsProperty = location.pathname.split('/')[2];
    return currentDocsProperty === repoName;
  }
```

rayangler commented on frontend/tests/unit/utils/mock-hapley-api.ts:
```diff
@@ -0,0 +1,12 @@
+import { GET_REPOS } from './mockData/repos';
+
+export const mockHapleyApi = (url: string) => {
+  const pathname = url.split('/')[1];
+  switch (pathname) {
+    case 'repos':
```
I like the pattern that we're establishing here to help ensure that we're using the correct/expected endpoint for a fetch. It'll be interesting to see how we build this out with other endpoints

rayangler reviewed (approved):
LGTM! Thanks for taking the time to get this foundation set up
